web/api/query.go

```diff
@@ -48,6 +68,12 @@ func (serv MetricsService) Query(w http.ResponseWriter, r *http.Request) {
 	expr := params.Get("expr")
 	asText := params.Get("asText")
 
+	timestamp, err := parseTimestampOrNow(params.Get("timestamp"))
+	if err != nil {
+		http.Error(w, fmt.Sprintf("invalid query timestamp %s", err), http.StatusBadRequest)
```

Comment:
Actually, I'm no longer sure about that. So far, I'm treating technical request errors differently from actual query language errors (like syntax errors). The former get an HTTP error status code and non-JSON text, while the latter get a JSON response and 200-OK (also for errors, where the error is only in the JSON then). This is because there are two code paths in the JS usually: one for handling successful AJAX responses, and one for the non-OK responses. The reason I think it's problematic to standardize on JSON for the non-OK responses is that they might be triggered at a point in the chain where we don't even have control over the output format yet (like in the HTTP framework, on the network, on a proxy server...). So doesn't it make more sense to keep that branch of control completely non-JSON?